param_sweep.py

```diff
@@ -0,0 +1,374 @@
+
+        desc_params = {i:v[0] for i, (v, _) in params.items() if len(v) == 1}
+        plt.title(f'Security with parameters {desc_params}')
+
+        for (j, i), label in np.ndenumerate(mat):
+            ax.text(i, j, round(label, 1), ha='center', va='center', color='white')
+
+        ax.set_xlabel(f'Parameter: {x_param}')
+        ax.set_ylabel(f'Parameter: {y_param}')
+
+        fig.savefig(file_name + '_gradient.png')
+        Logging.log('sweep', log_level, 'Saved the gradient graph to: %s',
+                    file_name + '_gradient.png')
+
+        if security_cutoff:
+            fig2, ax2 = plt.subplots(figsize=(20, 20), dpi=80)
+            binary_mat = mat >= security_cutoff
+            ax2.imshow(binary_mat)
+            ax2.set_xticks(np.arange(0, len(set(x)), 1))
+            ax2.set_yticks(np.arange(0, len(set(y)), 1))
+            ax2.set_xticklabels(sorted(list(set(x))))
+            ax2.set_yticklabels(sorted(list(set(y)), reverse=True))
+
+            for (j, i), label in np.ndenumerate(mat):
+                ax2.text(i,
+                         j,
+                         round(label, 1),
+                         ha='center',
+                         va='center',
+                         color='white' if label < security_cutoff else 'black')
+
+            for (j, i), label in np.ndenumerate(mat):
+                ax.text(i, j, round(label, 1), ha='center', va='center', color='white')
+            plt.title(f'Security with parameters {desc_params}'
+                      f' and cutoff at {security_cutoff}')
+
+            ax2.set_xlabel(f'Parameter: {x_param}')
+            ax2.set_ylabel(f'Parameter: {y_param}')
+
+            fig2.savefig(file_name + '_cutoff.png')
+            Logging.log('sweep', log_level, 'Saved the cutoff graph to: %s',
+                        file_name + '_cutoff.png')
+
+
+def PerformCalculation(input_params: tuple[int, float],
+                       results: dict,
+                       uniform_s: bool,
+                       fast: bool,
+                       log_level=None) -> None:
+    """
+    This function calls the lattice-estimator for a given set of input
+    parameters, and appends the output to the `results` dict. 
+    """
+    Logging.log('sweep', log_level, f'Running for n, q, Xe, Xs = {input_params}')
+    Xs_ = ND.UniformMod(input_params[2]) if uniform_s \
+        else ND.DiscreteGaussian(2**input_params[2])
+    f_ = LWE.estimate.rough if fast else LWE.estimate
+    params = LWE.Parameters(
+        n=int(input_params[0]),
+        q=int(input_params[1]),
+        Xe=ND.DiscreteGaussian(2**input_params[2]),
+        Xs=Xs_,
+        m=input_params[0],
+    )
+    print("params: ", params)
+    results[input_params] = security_level(params, f=f_)
+
+
+def security_level(params, f=LWE.estimate, *args, **kwargs):
+    res = f(params, *args, **kwargs)
+    return min([math.log(res_.rop, 2) for res_ in res.values()])
+
+
+if __name__ == "__main__":
```

Comment:
Is a command line script a good idea here? Wouldn't it be more flexible to call this from iPython? This would also simplify the code?

Reply:
That does make much more sense, I updated it the code to be callable from iPython / sage!